src/honeybadger/event_dispatch.cr

```diff
@@ -0,0 +1,102 @@
+require "./event"
+
+module Honeybadger
+  class EventDispatch
+    # Allow for a spike of events
+    @channel = Channel(Event).new(1 << 16)
+    @buffer = IO::Memory.new
+
+    getter? running = false
+
+    def send(event : Event)
+      ensure_running
+
+      select
+      when @channel.send event
+      else # if the buffer is full, we just skip
+      end
+    end
+
+    private def ensure_running
+      sync { start }
+    end
+
+    private def start
+      return if running?
+
+      @running = true
+      spawn run
+
+      at_exit do
+        # Give the consumer fiber a chance to buffer any further events
+        Fiber.yield
+        send @buffer.to_s
+      end
+    end
+
+    private def run
+      next_event = nil
+
+      # Main run loop
+      # 1. Buffer messages until either 60 seconds have passed or the buffer
+      #    exceeds 5MB
+      # 2. Flush the buffer to the Honeybadger API
+      # 3. Clear the buffer
+      loop do
+        buffering = true
+        wait_time = 60.seconds
+
+        while buffering
+          # If the previous iteration exceeded the buffer cap, we flushed it and
+          # now we need to place it into the fresh buffer
+          if next_event
+            @buffer.puts next_event
+            next_event = nil
+          end
+
+          started_waiting = Time.monotonic
+          select
+          when event = @channel.receive
+            # Limits for events endpoint: https://docs.honeybadger.io/api/reporting-events/#limits
+
+            json = event.to_json
+            # Max event size is 100KB
+            if json.bytesize <= 100 * 1024
+              # Maximum payload size is 5MB
+              if @buffer.bytesize + json.bytesize < 5 * 1024 * 1024
+                @buffer.puts json
+                next_event = nil
+              else
+                next_event = json
+                buffering = false
+              end
+            end
```

Comment:
Your comment about the potential for circular logic for logging is well noted, but the gotcha of silently failing when an event happens to be too large for the payload makes me feel even more uneasy.

Here's a possibility which mimic's what I think I've seen in other metrics/events systems. If the json bytesize is > 100kb, discard it and instead send an event which says that the system received an event too large to send. What do you think?

Reply:
What other systems do this? I think I'd only ever seen them dropped. I don't like that, though.

I like the idea of sending an event that says "there _should_ be an event here, but it was too big". Any ideas on how to ensure that there's sufficient information in _that_ event to debug the original event without also potentially exceeding the 100KB threshold?

Reply:
I don't have any specific memory of which system, you're asking too much! 😄 

I don't think it's possible to extract anything specific out of the event at this point. If the size check was instead within Event.new, a stack trace could be generated and captured...but that can be pretty expensive and I think I'd prefer this API to be return-as-soon-as-possible.

If the guard which checks the size of the event is in part of the json serialization, I think it might be possible to only fields from the NamedTuple on up to the byte size. Again, that's maybe more work than should be done in the same thread as the application. I'd have to tinker with it to know if it's possible to do that in the spawn. I'm not sure if it's worth it?

The honeybadger Ruby implementation emits a [fair amount of messages](https://github.com/honeybadger-io/honeybadger-ruby/blob/master/lib/honeybadger/events_worker.rb#L214). I think they're bypassing the Logger and just emitting the messages to stdout. At this time I don't actually see any code which limits either the byte size of the event or the byte size of the payload.

Reply:
> The Ruby client also defaults the queue size to [100k messages](https://github.com/honeybadger-io/honeybadger-ruby/blob/d6f76f77ae23c8bb2949cf09743f6bd72698bad2/lib/honeybadger/config/defaults.rb#L102), whereas I set the default to [64k](https://github.com/honeybadger-io/honeybadger-crystal/pull/37/files#diff-86036c4e621eb9d95ce36188a64efbbeaa56edaf94faefc670d5f84dbbafbc91R6). I'll bump that up to match, but I don't love that. Crystal channels [eagerly allocate their buffers](https://github.com/crystal-lang/crystal/issues/11572), so for most services that memory will just be sitting there allocated but unused. In my NATS client, I actually [overrode this behavior in a subclass of `Channel(T)`](https://github.com/jgaskins/nats/blob/80a6f50688efe4b1490787d59aaf1efa65dd94a8/src/nats.cr#L935-L952) due to the amount of memory used. I don't know if that would be useful here since there's only one of these (NATS clients can subscribe to thousands of subjects, so those channels _really_ shouldn't allocate hundreds of KB each that they don't need), so I'm open to suggestions there.

We started out with a lower queue size, but we had customers who were seeing messages being dropped regularly, so we bumped that up. I think 64k is a reasonable starting point.